src/library/scala/concurrent/Future.scala

```diff
@@ -787,11 +861,9 @@ object Future {
   // by just not ever using it itself. scala.concurrent
   // doesn't need to create defaultExecutionContext as
   // a side effect.
-  private[concurrent] object InternalCallbackExecutor extends ExecutionContext with BatchingExecutor {
-    override protected def unbatchedExecute(r: Runnable): Unit =
-      r.run()
-    override def reportFailure(t: Throwable): Unit =
-      throw new IllegalStateException("problem in scala.concurrent internal callback", t)
+  private[concurrent] final object InternalCallbackExecutor extends ExecutionContext with java.util.concurrent.Executor with BatchingExecutor {
+    override protected final def unbatchedExecute(r: Runnable): Unit = r.run()
+    override final def reportFailure(t: Throwable): Unit = ExecutionContext.defaultReporter(t)
```

Comment:
What's the motivation for no longer throwing the `IllegalStateException`? Are there internal callbacks which may throw exceptions now?

Reply:
The motivation for not throwing is that it could lead to broken code (callbacks not being executed). And I didn't want to throw a Fatal.

Reply:
Would it make sense to report the exception wrapped in an `IllegalStateException`, since this isn't supposed to happen? e.g.

```scala
override final def reportFailure(t: Throwable): Unit =
  ExecutionContext.defaultReporter(new IllegalStateException("problem in scala.concurrent internal callback", t)
```

either that or report both

```scala
override final def reportFailure(t: Throwable): Unit = {
  ExecutionContext.defaultReporter(new IllegalStateException("problem in scala.concurrent internal callback"/*, t*/)
  ExecutionContext.defaultReporter(t)
}
```

Reply:
@NthPortal Great idea!